# Hacker News Posts

# <p style="background-color:	#D1D1D1;font-family:newtimeroman;color:#141414;font-size:120%;text-align:center;border-radius:10px 5px;"> Table of contents</p>

<a id="toc"></a>

* [   About Dataset](#0)
     

* [1) LIBRARIES NEEDED IN THE STUDY](#1)

* [2) DATA](#2)
       
* [3) EXPLORATORY DATA ANALYSIS](#3)


## About Dataset

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.  
The dataset can be found [here](https://www.kaggle.com/hacker-news/hacker-news-posts)  
Data description

| Data | Description |
| --- | --- |
| id | The unique identifier from Hacker News for the post |
| title | The title of the post |
| url | The URL that the posts links to, if it the post has a URL |
| num_points | The number of points the post acquired |
| num_comments | The number of comments that were made on the post |
| author | The username of the person who submitted the post |
| created_at | The date and time at which the post was submitted |

## 1) Import libraries

In [5]:
import datetime as dt
import pandas as pd

## 2) Data

In [18]:
import csv

f = open('hacker_news.csv')
hn = list(csv.reader(f))
hn[:5]
header = hn[0]

In [19]:
hn

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
 

In [20]:
header

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In this dataset, I'll focus on exploring 2 types of posts sush as `Ask HN` and `Show Hn`

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?

## 3) EDA 

### Extracting Ask HN and Show HN Posts

First, we'll identify posts that begin with either `Ask HN` or `Show HN` and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [23]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    if title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
18939


In [24]:
print('Ask posts count:' + " " + str(len(ask_posts)))
print('\n')
print('Show posts count:' + " " + str(len(show_posts)))
print('\n')
print('Other posts count:' + " " + str(len(other_posts))) 

Ask posts count: 1744


Show posts count: 1162


Other posts count: 18939


## Calculating the Average Number of Comments for Ask HN and Show HN Posts

### Average Comments per Post - 'Ask HN' 

In [26]:
header

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In [28]:
total_ask_comments = 0
for post in ask_posts:
    total_ask_comments += int(post[4])
avg_ask_comments = total_ask_comments / len(ask_posts)

In [29]:
avg_ask_comments

14.038417431192661

In [32]:
print(round(avg_ask_comments,2), 'comments per post on average.')

14.04 comments per post on average.


### Average Comments per Post - 'Show HN' 

In [33]:
total_show_comments = 0
for post in show_posts:
    total_show_comments += int(post[4])
avg_show_comments = total_show_comments / len(show_posts)   

In [34]:
avg_show_comments

10.31669535283993

In [35]:
print(round(avg_show_comments,2), 'comments per post on average.')

10.32 comments per post on average.


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

## Finding the Amount of Ask Posts and Comments by Hour Created

Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [36]:
# List comments in ask_post by datetime
result_list = []
for post in ask_posts:
    result_list.append([post[6],post[4]])

In [37]:
result_list

[['8/16/2016 9:55', '6'],
 ['11/22/2015 13:43', '29'],
 ['5/2/2016 10:14', '1'],
 ['8/2/2016 14:20', '3'],
 ['10/15/2015 16:38', '17'],
 ['9/26/2015 23:23', '1'],
 ['4/22/2016 12:24', '4'],
 ['11/16/2015 9:22', '1'],
 ['2/24/2016 17:57', '1'],
 ['6/4/2016 17:17', '2'],
 ['9/19/2015 17:04', '7'],
 ['9/22/2015 13:16', '1'],
 ['6/21/2016 15:45', '1'],
 ['1/13/2016 21:17', '4'],
 ['10/4/2015 21:27', '4'],
 ['1/25/2016 20:27', '2'],
 ['10/27/2015 2:47', '3'],
 ['1/19/2016 12:01', '1'],
 ['3/22/2016 2:05', '22'],
 ['9/8/2015 14:04', '2'],
 ['8/28/2016 18:06', '2'],
 ['7/20/2016 13:44', '7'],
 ['9/12/2016 16:52', '7'],
 ['2/29/2016 17:52', '3'],
 ['4/18/2016 15:28', '6'],
 ['12/28/2015 14:38', '2'],
 ['4/4/2016 3:34', '1'],
 ['1/15/2016 21:47', '3'],
 ['11/19/2015 5:33', '29'],
 ['12/20/2015 3:59', '2'],
 ['10/15/2015 21:34', '20'],
 ['2/26/2016 19:20', '3'],
 ['8/2/2016 18:00', '3'],
 ['2/28/2016 1:24', '33'],
 ['1/13/2016 9:12', '5'],
 ['5/6/2016 1:14', '4'],
 ['6/23/2016 13:59', '7'],
 ['4

In [44]:
comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    date = each_row[0]
    comment = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1

comments_by_hour
        

{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

## Calculating the Average Number of Comments for Ask HN Posts by Hour

In [45]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour = []
for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr]/ counts_by_hour[hr]])

In [49]:
avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

## Sorting and Printing Values from a List of Lists

In [50]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

sorted_swap

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [51]:
# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.